libsql-wal/src/wal.rs

```diff
@@ -0,0 +1,303 @@
+        Ok(invalidate_cache)
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn end_read_txn(&mut self) {
+        match self.tx.take() {
+            Some(Transaction::Read(tx)) => tx,
+            Some(Transaction::Write(tx)) => tx.downgrade(),
+            None => return,
+        };
+
+        tracing::trace!("end read tx");
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn find_frame(
+        &mut self,
+        page_no: std::num::NonZeroU32,
+    ) -> libsql_sys::wal::Result<Option<std::num::NonZeroU32>> {
+        tracing::trace!(page_no, "find frame");
+        // this is a trick: we defer the frame read to the `read_frame` method. The read_frame
+        // method will read from the journal if the page exist, or from the db_file if it doesn't
+        Ok(Some(page_no))
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn read_frame(
+        &mut self,
+        page_no: std::num::NonZeroU32,
+        buffer: &mut [u8],
+    ) -> libsql_sys::wal::Result<()> {
+        tracing::trace!(page_no, "reading frame");
+        let tx = self.tx.as_mut().unwrap();
+        self.shared
+            .read_frame(tx, page_no.get(), buffer)
+            .map_err(Into::into)?;
+        Ok(())
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn db_size(&self) -> u32 {
+        let db_size = match self.tx.as_ref() {
+            Some(tx) => tx.db_size,
+            None => 0,
+        };
+        tracing::trace!(db_size, "db_size");
+        db_size
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn begin_write_txn(&mut self) -> libsql_sys::wal::Result<()> {
+        tracing::trace!("begin write");
+        match self.tx.as_mut() {
+            Some(tx) => {
+                self.shared.upgrade(tx).map_err(Into::into)?;
+                tracing::debug!("write lock acquired");
+            }
+            None => todo!("should acquire read txn first"),
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn end_write_txn(&mut self) -> libsql_sys::wal::Result<()> {
+        tracing::trace!("end write");
+        match self.tx.take() {
+            Some(Transaction::Write(tx)) => {
+                self.last_read_frame_no = Some(tx.next_frame_no - 1);
+                self.tx = Some(Transaction::Read(tx.downgrade()));
+            }
+            other => {
+                self.tx = other;
+            }
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn undo<U: libsql_sys::wal::UndoHandler>(
+        &mut self,
+        handler: Option<&mut U>,
+    ) -> libsql_sys::wal::Result<()> {
+        match self.tx {
+            Some(Transaction::Write(ref mut tx)) => {
+                if tx.is_commited() {
+                    return Ok(());
+                }
```

Comment:
shouldn't this return an error since you are calling undo on a committed transaction

Reply:
That's what I initially thought, but SQLite can call undo after a commit, and that's noop. I found that with fault injection.